[package] folly/2020.08.10.00: undefined reference to `folly::SharedMutexImpl<false, void, std::atomic, false, false>::SharedMutexImpl()'

### Package and Environment Details (include every applicable attribute)
  * Package Name/Version: **folly/2020.08.10.00**
  * Operating System+version: **Linux Ubuntu 18.04**
  * Compiler+version: **GCC 9**
  * Conan version: **conan 1.30.2**
  * Python version: **Python 3.7.4**


### Conan profile (output of `conan profile show default` or `conan profile show <profile>` if custom profile is in use)
```
[settings]
os=Linux
os_build=Linux
arch=x86_64
arch_build=x86_64
compiler=gcc
compiler.version=9
compiler.libcxx=libstdc++
build_type=Release
[options]
[build_requires]
[env]
```


### Steps to reproduce (Include if Applicable)

Link using `folly/2020.08.10.00`.  You get undefined symbols like
```
undefined reference to `folly::SharedMutexImpl<false, void, std::atomic, false, false>::SharedMutexImpl()'
```

The reason is because of link ordering within Folly.  The recipe uses `collect_libs()`.  Replacing `collect_libs()`  library ordering fixes the problem:
```
        self.cpp_info.components["libfolly"].libs = [
            "folly_exception_counter",
            "folly_exception_tracer",
            "folly_exception_tracer_base",
            "folly_test_util",
            "follybenchmark",
            "folly"
        ]
```


### Logs (Include/Attach if Applicable)
<details><summary>Click to expand log</summary>

```
/usr/bin/ld: /home/pfraval/.conan/data/folly/2020.08.10.00/_/_/package/d474cd0f413a68e19aaa3dc823ff9d479ab99b7c/lib/libfolly_exception_tracer.a(ExceptionTracerLib.cpp.o): in function `folly::Synchronized<std::vector<void (*)(void*, std::type_info*, void (**)(void*)), std::allocator<void (*)(void*, std::type_info*, void (**)(void*))> >, folly::SharedMutexImpl<false, void, std::atomic, false, false> >::Synchronized()':
/home/conan/w/cci_PR-2363/.conan/data/folly/2020.08.10.00/_/_/build/d474cd0f413a68e19aaa3dc823ff9d479ab99b7c/source_subfolder/folly/Synchronized.h:513: undefined reference to `folly::SharedMutexImpl<false, void, std::atomic, false, false>::SharedMutexImpl()'
/usr/bin/ld: /home/pfraval/.conan/data/folly/2020.08.10.00/_/_/package/d474cd0f413a68e19aaa3dc823ff9d479ab99b7c/lib/libfolly_exception_tracer.a(ExceptionTracerLib.cpp.o): in function `folly::Synchronized<std::vector<void (*)(void*), std::allocator<void (*)(void*)> >, folly::SharedMutexImpl<false, void, std::atomic, false, false> >::Synchronized()':
/home/conan/w/cci_PR-2363/.conan/data/folly/2020.08.10.00/_/_/build/d474cd0f413a68e19aaa3dc823ff9d479ab99b7c/source_subfolder/folly/Synchronized.h:513: undefined reference to `folly::SharedMutexImpl<false, void, std::atomic, false, false>::SharedMutexImpl()'
/usr/bin/ld: /home/pfraval/.conan/data/folly/2020.08.10.00/_/_/package/d474cd0f413a68e19aaa3dc823ff9d479ab99b7c/lib/libfolly_exception_tracer.a(ExceptionTracerLib.cpp.o): in function `folly::Synchronized<std::vector<void (*)(), std::allocator<void (*)()> >, folly::SharedMutexImpl<false, void, std::atomic, false, false> >::Synchronized()':
/home/conan/w/cci_PR-2363/.conan/data/folly/2020.08.10.00/_/_/build/d474cd0f413a68e19aaa3dc823ff9d479ab99b7c/source_subfolder/folly/Synchronized.h:513: undefined reference to `folly::SharedMutexImpl<false, void, std::atomic, false, false>::SharedMutexImpl()'
/usr/bin/ld: /home/pfraval/.conan/data/folly/2020.08.10.00/_/_/package/d474cd0f413a68e19aaa3dc823ff9d479ab99b7c/lib/libfolly_exception_tracer.a(ExceptionTracerLib.cpp.o): in function `folly::Synchronized<std::vector<void (*)(std::__exception_ptr::exception_ptr), std::allocator<void (*)(std::__exception_ptr::exception_ptr)> >, folly::SharedMutexImpl<false, void, std::atomic, false, false> >::Synchronized()':
/home/conan/w/cci_PR-2363/.conan/data/folly/2020.08.10.00/_/_/build/d474cd0f413a68e19aaa3dc823ff9d479ab99b7c/source_subfolder/folly/Synchronized.h:513: undefined reference to `folly::SharedMutexImpl<false, void, std::atomic, false, false>::SharedMutexImpl()'
/usr/bin/ld: /home/pfraval/.conan/data/folly/2020.08.10.00/_/_/package/d474cd0f413a68e19aaa3dc823ff9d479ab99b7c/lib/libfolly_exception_tracer.a(ExceptionTracerLib.cpp.o): in function `folly::detail::LockTraitsImpl<folly::SharedMutexImpl<false, void, std::atomic, false, false>, (folly::detail::MutexLevel)1, false>::unlock_shared(folly::SharedMutexImpl<false, void, std::atomic, false, false>&)':
/home/conan/w/cci_PR-2363/.conan/data/folly/2020.08.10.00/_/_/build/d474cd0f413a68e19aaa3dc823ff9d479ab99b7c/source_subfolder/folly/LockTraits.h:164: undefined reference to `folly::SharedMutexImpl<false, void, std::atomic, false, false>::unlock_shared()'
/usr/bin/ld: /home/pfraval/.conan/data/folly/2020.08.10.00/_/_/package/d474cd0f413a68e19aaa3dc823ff9d479ab99b7c/lib/libfolly_exception_tracer.a(ExceptionTracerLib.cpp.o): in function `folly::detail::LockTraitsImpl<folly::SharedMutexImpl<false, void, std::atomic, false, false>, (folly::detail::MutexLevel)0, false>::unlock(folly::SharedMutexImpl<false, void, std::atomic, false, false>&)':
/home/conan/w/cci_PR-2363/.conan/data/folly/2020.08.10.00/_/_/build/d474cd0f413a68e19aaa3dc823ff9d479ab99b7c/source_subfolder/folly/LockTraits.h:131: undefined reference to `folly::SharedMutexImpl<false, void, std::atomic, false, false>::unlock()'
/usr/bin/ld: /home/pfraval/.conan/data/folly/2020.08.10.00/_/_/package/d474cd0f413a68e19aaa3dc823ff9d479ab99b7c/lib/libfolly_exception_tracer.a(ExceptionTracerLib.cpp.o): in function `folly::detail::LockTraitsImpl<folly::SharedMutexImpl<false, void, std::atomic, false, false>, (folly::detail::MutexLevel)0, false>::lock(folly::SharedMutexImpl<false, void, std::atomic, false, false>&)':
/home/conan/w/cci_PR-2363/.conan/data/folly/2020.08.10.00/_/_/build/d474cd0f413a68e19aaa3dc823ff9d479ab99b7c/source_subfolder/folly/LockTraits.h:124: undefined reference to `folly::SharedMutexImpl<false, void, std::atomic, false, false>::lock()'
/usr/bin/ld: /home/pfraval/.conan/data/folly/2020.08.10.00/_/_/package/d474cd0f413a68e19aaa3dc823ff9d479ab99b7c/lib/libfolly_exception_tracer.a(ExceptionTracerLib.cpp.o): in function `folly::detail::LockTraitsImpl<folly::SharedMutexImpl<false, void, std::atomic, false, false>, (folly::detail::MutexLevel)1, false>::lock_shared(folly::SharedMutexImpl<false, void, std::atomic, false, false>&)':
/home/conan/w/cci_PR-2363/.conan/data/folly/2020.08.10.00/_/_/build/d474cd0f413a68e19aaa3dc823ff9d479ab99b7c/source_subfolder/folly/LockTraits.h:157: undefined reference to `folly::SharedMutexImpl<false, void, std::atomic, false, false>::lock_shared()'
collect2: error: ld returned 1 exit status
make[2]: *** [tw/service/test/CMakeFiles/service-test-runner.dir/build.make:555: bin/service-test-runner] Error 1
make[2]: Leaving directory '/home/pfraval/dev/ifi/base/cmake-build-debug'
make[1]: *** [CMakeFiles/Makefile2:1706: tw/service/test/CMakeFiles/service-test-runner.dir/all] Error 2
make[1]: Leaving directory '/home/pfraval/dev/ifi/base/cmake-build-debug'
make: *** [Makefile:160: all] Error 2
```

</details>

I couldn't figure out how to replicate the original problem, but I could replicate this other undefined reference which is very similar:
```
/usr/bin/ld: /home/eric/.conan/data/folly/2020.08.10.00/_/_/package/b8af69894efc20cc2a5098146dca4122c1f071fe/lib/libfolly_exception_tracer_base.a(ExceptionTracer.cpp.o): in function `folly::exception_tracer::printExceptionInfo(std::ostream&, folly::exception_tracer::ExceptionInfo const&, int)':
ExceptionTracer.cpp:(.text+0x23c): undefined reference to `folly::symbolizer::Symbolizer::Symbolizer(folly::symbolizer::ElfCacheBase*, folly::symbolizer::LocationInfoMode, unsigned long)'
/usr/bin/ld: ExceptionTracer.cpp:(.text+0x25e): undefined reference to `folly::symbolizer::Symbolizer::symbolize(folly::Range<unsigned long const*>, folly::Range<folly::symbolizer::SymbolizedFrame*>)'
/usr/bin/ld: ExceptionTracer.cpp:(.text+0x271): undefined reference to `folly::symbolizer::OStreamSymbolizePrinter::OStreamSymbolizePrinter(std::ostream&, int)'
/usr/bin/ld: ExceptionTracer.cpp:(.text+0x281): undefined reference to `folly::symbolizer::SymbolizePrinter::println(folly::symbolizer::SymbolizedFrame const*, unsigned long)'
/usr/bin/ld: /home/eric/.conan/data/folly/2020.08.10.00/_/_/package/b8af69894efc20cc2a5098146dca4122c1f071fe/lib/libfolly_exception_tracer_base.a(ExceptionTracer.cpp.o): in function `folly::exception_tracer::printExceptionInfo(std::ostream&, folly::exception_tracer::ExceptionInfo const&, int) [clone .cold]':
ExceptionTracer.cpp:(.text.unlikely+0xc3): undefined reference to `folly::exceptionStr(std::exception const&)'
/usr/bin/ld: /home/eric/.conan/data/folly/2020.08.10.00/_/_/package/b8af69894efc20cc2a5098146dca4122c1f071fe/lib/libfolly_exception_tracer_base.a(StackTrace.cpp.o): in function `folly::exception_tracer::StackTraceStack::pushCurrent()':
StackTrace.cpp:(.text+0x75): undefined reference to `folly::symbolizer::getStackTrace(unsigned long*, unsigned long)'
collect2: error: ld returned 1 exit status
```

By adding the following to `test_package.cpp`:
```
#include <folly/experimental/exception_tracer/ExceptionTracer.h>
```
And then in `main()`:
```
    folly::exception_tracer::installHandlers();
```
Tested your PR - it fixes the problem.

--- a/recipes/folly/all/conanfile.py
+++ b/recipes/folly/all/conanfile.py
@@ -189,7 +189,7 @@ class FollyConan(ConanFile):
                 self.cpp_info.components["libfolly"].defines.append("FOLLY_HAVE_ELF")
 
         elif self.settings.os == "Windows":
-            self.cpp_info.components["libfolly"].system_libs.extend(["ws2_32", "Iphlpapi", "Crypt32"])
+            self.cpp_info.components["libfolly"].system_libs.extend(["ws2_32", "iphlpapi", "crypt32"])
 
         if (self.settings.os == "Linux" and self.settings.compiler == "clang" and
             self.settings.compiler.libcxx == "libstdc++") or \

--- a/recipes/folly/all/conanfile.py
+++ b/recipes/folly/all/conanfile.py
@@ -3,7 +3,8 @@ from conans import ConanFile, CMake, tools
 from conans.tools import Version
 from conans.errors import ConanInvalidConfiguration
 
-required_conan_version = ">=1.28.0"
+required_conan_version = ">=1.32.0"
+
 
 class FollyConan(ConanFile):
     name = "folly"
@@ -44,74 +45,61 @@ class FollyConan(ConanFile):
     def configure(self):
         if self.options.shared:
             del self.options.fPIC
-        if self.settings.get_safe("compiler.cppstd"):
+        if self.settings.compiler.get_safe("cppstd"):
             tools.check_min_cppstd(self, self._minimum_cpp_standard)
-        min_version = self._minimum_compilers_version.get(
-            str(self.settings.compiler))
+        min_version = self._minimum_compilers_version.get(str(self.settings.compiler))
         if not min_version:
-            self.output.warn("{} recipe lacks information about the {} compiler support.".format(
-                self.name, self.settings.compiler))
+            self.output.warn("{} recipe lacks information about the {} compiler support.".format(self.name, self.settings.compiler))
         else:
             if tools.Version(self.settings.compiler.version) < min_version:
-                raise ConanInvalidConfiguration("{} requires C++14 support. The current compiler {} {} does not support it.".format(
-                    self.name, self.settings.compiler, self.settings.compiler.version))
+                raise ConanInvalidConfiguration("{} requires C++{} support. The current compiler {} {} does not support it.".format(
+                    self.name, self._minimum_cpp_standard, self.settings.compiler, self.settings.compiler.version))
 
         if self.settings.os == "Windows" and self.settings.arch != "x86_64":
-            raise ConanInvalidConfiguration("Folly requires a 64bit target architecture")
-        elif self.settings.os == "Windows" and self.settings.compiler == "Visual Studio" and \
-                "MT" in self.settings.compiler.runtime:
-            raise ConanInvalidConfiguration("Folly could not be build with runtime MT")
-        elif self.settings.os == "Macos" and self.options.shared:
-            raise ConanInvalidConfiguration("Folly could not be built by apple-clang as shared library")
-        elif self.settings.os == "Windows" and self.options.shared:
-            raise ConanInvalidConfiguration("Folly could not be built on Windows as a shared library")
-        elif Version(self.version) >= "2020.08.10.00" and self.settings.compiler == "Visual Studio" and \
-                not self.options.shared:
-            raise ConanInvalidConfiguration("Folly could not be built on Windows as a static library")
-        elif Version(self.version) >= "2020.08.10.00" and self.settings.compiler == "clang" and \
-                self.options.shared:
+            raise ConanInvalidConfiguration("Folly requires a 64bit target architecture on Windows")
+
+        if self.settings.os in ["Macos", "Windows"] and self.options.shared:
+            raise ConanInvalidConfiguration("Folly could not be built on {} as shared library".format(self.settings.os))
+
+        if self.version == "2020.08.10.00" and self.settings.compiler == "clang" and self.options.shared:
             raise ConanInvalidConfiguration("Folly could not be built by clang as a shared library")
-
-        self._strip_options_requirements()
-
-    def _strip_options_requirements(self):
-        self.options["boost"].header_only = False
-        for boost_comp in self._required_boost_components:
-            setattr(self.options["boost"], "without_{}".format(boost_comp), False)
-
-    @property
-    def _required_boost_components(self):
-        return ["context", "filesystem", "program_options", "regex", "system", "thread"]
 
     # Freeze max. CMake version at 3.16.2 to fix the Linux build
     def build_requirements(self):
         self.build_requires("cmake/3.16.2")
 
     def requirements(self):
-        self.requires("boost/1.74.0")
+        self.requires("boost/1.75.0")
         self.requires("bzip2/1.0.8")
         self.requires("double-conversion/3.1.5")
         self.requires("gflags/2.2.2")
         self.requires("glog/0.4.0")
         self.requires("libevent/2.1.12")
-        self.requires("lz4/1.9.2")
+        self.requires("lz4/1.9.3")
         self.requires("openssl/1.1.1i")
         self.requires("snappy/1.1.8")
         self.requires("zlib/1.2.11")
-        self.requires("zstd/1.4.5")
-        if Version(self.version) >= "2019.01.01.00":
-            self.requires("libdwarf/20191104")
-            self.requires("libsodium/1.0.18")
-            self.requires("xz_utils/5.2.4")
-            if self.settings.os == "Linux":
-                self.requires("libiberty/9.1.0")
-                self.requires("libunwind/1.5.0")
+        self.requires("zstd/1.4.8")
+        self.requires("libdwarf/20191104")
+        self.requires("libsodium/1.0.18")
+        self.requires("xz_utils/5.2.5")
+        # FIXME: Causing compilation issues on clang: self.requires("jemalloc/5.2.1")
+        if self.settings.os == "Linux":
+            self.requires("libiberty/9.1.0")
+            self.requires("libunwind/1.5.0")
         if Version(self.version) >= "2020.08.10.00":
             self.requires("fmt/7.0.3")
 
-    def _validate_dependency_graph(self):
+    @property
+    def _required_boost_components(self):
+        return ["context", "filesystem", "program_options", "regex", "system", "thread"]
+
+    def validate(self):
+        if self.options["boost"].header_only:
+            raise ConanInvalidConfiguration("Folly could not be built with a header only Boost")
+
         miss_boost_required_comp = any(getattr(self.options["boost"], "without_{}".format(boost_comp), True) for boost_comp in self._required_boost_components)
-        if self.options["boost"].header_only or miss_boost_required_comp:
+        if miss_boost_required_comp:
             raise ConanInvalidConfiguration("Folly requires these boost components: {}".format(", ".join(self._required_boost_components)))
 
     def source(self):
@@ -123,8 +111,7 @@ class FollyConan(ConanFile):
         if not self._cmake:
             self._cmake = CMake(self)
             self._cmake.definitions["CMAKE_POSITION_INDEPENDENT_CODE"] = self.options.get_safe("fPIC", True)
-            self._cmake.definitions["CXX_STD"] = self.settings.get_safe(
-                "compiler.cppstd") or "c++14"
+            self._cmake.definitions["CXX_STD"] = self.settings.compiler.get_safe("cppstd") or "c++14"
             if self.settings.compiler == "Visual Studio":
                 self._cmake.definitions["MSVC_ENABLE_ALL_WARNINGS"] = False
                 self._cmake.definitions["MSVC_USE_STATIC_RUNTIME"] = "MT" in self.settings.compiler.runtime
@@ -132,15 +119,13 @@ class FollyConan(ConanFile):
         return self._cmake
 
     def build(self):
-        self._validate_dependency_graph()
         for patch in self.conan_data["patches"][self.version]:
             tools.patch(**patch)
         cmake = self._configure_cmake()
         cmake.build()
 
     def package(self):
-        self.copy(pattern="LICENSE", dst="licenses",
-                  src=self._source_subfolder)
+        self.copy(pattern="LICENSE", dst="licenses", src=self._source_subfolder)
         cmake = self._configure_cmake()
         cmake.install()
         tools.rmdir(os.path.join(self.package_folder, "lib", "cmake"))
@@ -155,7 +140,29 @@ class FollyConan(ConanFile):
         self.cpp_info.components["libfolly"].names["cmake_find_package"] = "folly"
         self.cpp_info.components["libfolly"].names["cmake_find_package_multi"] = "folly"
         self.cpp_info.components["libfolly"].names["pkg_config"] = "libfolly"
-        self.cpp_info.components["libfolly"].libs = tools.collect_libs(self)
+
+        if Version(self.version) == "2019.10.21.00":
+            self.cpp_info.components["libfolly"].libs = [
+                "follybenchmark",
+                "folly_test_util",
+                "folly"
+            ]
+        if Version(self.version) == "2020.08.10.00":
+            if self.settings.os == "Linux":
+                self.cpp_info.components["libfolly"].libs = [
+                    "folly_exception_counter",
+                    "folly_exception_tracer",
+                    "folly_exception_tracer_base",
+                    "folly_test_util",
+                    "follybenchmark",
+                    "folly"
+                ]
+            else:
+                self.cpp_info.components["libfolly"].libs = [
+                    "folly_test_util",
+                    "follybenchmark",
+                    "folly"
+                ]
         self.cpp_info.components["libfolly"].requires = [
             "boost::boost",
             "bzip2::bzip2",
@@ -167,25 +174,23 @@ class FollyConan(ConanFile):
             "openssl::openssl",
             "snappy::snappy",
             "zlib::zlib",
-            "zstd::zstd"
+            "zstd::zstd",
+            "libdwarf::libdwarf",
+            "libsodium::libsodium",
+            "xz_utils::xz_utils"
         ]
-        if Version(self.version) >= "2019.01.01.00":
-            self.cpp_info.components["libfolly"].requires.extend([
-                "libdwarf::libdwarf",
-                "libsodium::libsodium",
-                "xz_utils::xz_utils"
-            ])
-            if self.settings.os == "Linux":
-                self.cpp_info.components["libfolly"].requires.extend([
-                    "libiberty::libiberty",
-                    "libunwind::libunwind"
-                ])
+        if self.settings.os == "Linux":
+            self.cpp_info.components["libfolly"].requires.extend(["libiberty::libiberty", "libunwind::libunwind"])
+            self.cpp_info.components["libfolly"].system_libs.extend(["pthread", "dl", "rt"])
+
         if Version(self.version) >= "2020.08.10.00":
             self.cpp_info.components["libfolly"].requires.append("fmt::fmt")
-        if self.settings.os == "Linux":
-            self.cpp_info.components["libfolly"].system_libs.extend(["pthread", "dl", "rt"])
+            if self.settings.os == "Linux":
+                self.cpp_info.components["libfolly"].defines.append("FOLLY_HAVE_ELF")
+
         elif self.settings.os == "Windows":
             self.cpp_info.components["libfolly"].system_libs.extend(["ws2_32", "Iphlpapi", "Crypt32"])
+
         if (self.settings.os == "Linux" and self.settings.compiler == "clang" and
             self.settings.compiler.libcxx == "libstdc++") or \
            (self.settings.os == "Macos" and self.settings.compiler == "apple-clang" and

--- a/recipes/folly/all/test_package/test_package.cpp
+++ b/recipes/folly/all/test_package/test_package.cpp
@@ -6,6 +6,9 @@
 #include <folly/executors/ThreadedExecutor.h>
 #include <folly/Uri.h>
 #include <folly/FBString.h>
+#if FOLLY_HAVE_ELF
+#include <folly/experimental/symbolizer/Elf.h>
+#endif
 
 static void print_uri(const folly::fbstring& value) {
     const folly::Uri uri(value);
@@ -20,5 +23,8 @@ int main() {
     folly::Future<folly::Unit> unit = std::move(future).thenValue(print_uri);
     promise.setValue("https://github.com/bincrafters");
     std::move(unit).get();
+#if FOLLY_HAVE_ELF
+    folly::symbolizer::ElfFile elffile;
+#endif
     return EXIT_SUCCESS;
 }

